File: evaluation/typescript/src/evaluation.ts
import { EVALUATOR_ERRORS } from './errors';
import { Clause } from './proto/feature/clause_pb';
import { Evaluation, UserEvaluations } from './proto/feature/evaluation_pb';
import { Feature } from './proto/feature/feature_pb';
import { Reason } from './proto/feature/reason_pb';
import { SegmentUser } from './proto/feature/segment_pb';
import { Variation } from './proto/feature/variation_pb';
import { User } from './proto/user/user_pb';
import { RuleEvaluator } from './ruleEvaluator';
import { StrategyEvaluator } from './strategyEvaluator';
import { NewUserEvaluations, UserEvaluationsID } from './userEvaluation';
import { createReason } from './modelFactory';

const SECONDS_TO_RE_EVALUATE_ALL = 30 * 24 * 60 * 60; // 30 days
const SECONDS_FOR_ADJUSTMENT = 10; // 10 seconds

export function EvaluationID(featureID: string, featureVersion: number, userID: string): string {
  return `${featureID}:${featureVersion}:${userID}`;
}

class Evaluator {
  private ruleEvaluator: RuleEvaluator;
  private strategyEvaluator: StrategyEvaluator;

  constructor() {
    this.ruleEvaluator = new RuleEvaluator();
    this.strategyEvaluator = new StrategyEvaluator();
  }

  async evaluateFeatures(
    features: Feature[],
    user: User,
    mapSegmentUsers: Map<string, SegmentUser[]>,
    targetTag: string,
  ): Promise<UserEvaluations> {
    return this.evaluate(features, user, mapSegmentUsers, false, targetTag);
  }

  // GetPrerequisiteDownwards gets the features specified as prerequisite by the targetFeatures.
  getPrerequisiteDownwards(targetFeatures: Feature[], allFeatures: Feature[]): Feature[] {
    const allFeaturesMap = new Map<string, Feature>();
    for (const feature of allFeatures) {
      allFeaturesMap.set(feature.getId(), feature);
    }
    const dependedOnTargets = getFeaturesDependedOnTargets(targetFeatures, allFeaturesMap);
    return Object.values(dependedOnTargets);
  }

  evaluateFeaturesByEvaluatedAt(
    features: Feature[],
    user: User,
    mapSegmentUsers: Map<string, SegmentUser[]>,
    prevUEID: string,
    evaluatedAt: number,
    userAttributesUpdated: boolean,
    targetTag: string,
  ): UserEvaluations {
    if (prevUEID === '') {
      return this.evaluate(features, user, mapSegmentUsers, true, targetTag);
    }

    const now = Math.floor(Date.now() / 1000);
    if (evaluatedAt < now - SECONDS_TO_RE_EVALUATE_ALL) {
      return this.evaluate(features, user, mapSegmentUsers, true, targetTag);
    }

    const adjustedEvalAt = evaluatedAt - SECONDS_FOR_ADJUSTMENT;
    const updatedFeatures = features.filter(
      (feature) =>
        feature.getUpdatedAt() > adjustedEvalAt ||
        (userAttributesUpdated && feature.getRulesList().length > 0),
    );

    // If the UserEvaluationsID has changed, but both User Attributes and Feature Flags have not been updated,
    // it is considered unusual and a force update should be performed.
    if (updatedFeatures.length === 0) {
      return this.evaluate(features, user, mapSegmentUsers, true, targetTag);
    }

    const evalTargets = this.getEvalFeatures(updatedFeatures, features);
    return this.evaluate(evalTargets, user, mapSegmentUsers, false, targetTag);
  }

  private evaluate(
    features: Feature[],
    user: User,
    mapSegmentUsers: Map<string, SegmentUser[]>,
    forceUpdate: boolean,
    targetTag: string,
  ): UserEvaluations {
    const flagVariations: { [key: string]: string } = {};
    // fs need to be sorted in order from upstream to downstream.
    const sortedFeatures = topologicalSort(features);

    const evaluations: Evaluation[] = [];
    const archivedIDs: string[] = [];

    for (const feature of sortedFeatures) {
      if (feature.getArchived()) {
        // To keep response size small, the feature flags archived long time ago are excluded.
        if (!this.isArchivedBeforeLastThirtyDays(feature)) {
          archivedIDs.push(feature.getId());
        }
        continue;
      }

      const segmentUsers = this.listSegmentIDs(feature).flatMap(
        (id) => mapSegmentUsers.get(id) || [],
      );

      const [reason, variation] = this.assignUser(feature, user, segmentUsers, flagVariations);
      // VariationId is used to check if prerequisite flag's result is what user expects it to be.
      flagVariations[feature.getId()] = variation.getId();
      // When the tag is set in the request,
      // it will return only the evaluations of flags that match the tag configured on the dashboard.
      // When empty, it will return all the evaluations of the flags in the environment.
      if (targetTag !== '' && !this.tagExist(feature.getTagsList(), targetTag)) {
        continue;
      }

      const evaluationID = EvaluationID(feature.getId(), feature.getVersion(), user.getId());
      const evaluation = new Evaluation();
      evaluation.setId(evaluationID);
      evaluation.setFeatureId(feature.getId());
      evaluation.setFeatureVersion(feature.getVersion());
      evaluation.setUserId(user.getId());
      evaluation.setVariationId(variation.getId());
      evaluation.setVariationName(variation.getName());
      evaluation.setVariationValue(variation.getValue());
      // Deprecated
      // FIXME: Remove the Variation when is no longer being used.
      // For security reasons, we should remove the variation description.
      // We copy the variation object to avoid race conditions when removing
      // the description directly from the `variation`
      const copyVariation = new Variation();
      copyVariation.setId(variation.getId());
      copyVariation.setName(variation.getName());
      copyVariation.setValue(variation.getValue());
      evaluation.setVariation(copyVariation);
      evaluation.setReason(reason);

      evaluations.push(evaluation);
    }

    const id = UserEvaluationsID(
      user.getId(),
      arrayToRecord(user.getDataMap().toArray()),
      features,
    );
    const userEvaluations = NewUserEvaluations(id, evaluations, archivedIDs, forceUpdate);
    return userEvaluations;
  }

  private tagExist(tags: string[], target: string): boolean {
    return tags.includes(target);
  }

  /*
  IsArchivedBeforeLastThirtyDays returns a bool value
  indicating whether the feature flag was archived within the last thirty days.
  */
  private isArchivedBeforeLastThirtyDays(feature: Feature): boolean {
    if (!feature.getArchived()) {
      return false;
    }
    const now = Math.floor(Date.now() / 1000);
    return feature.getUpdatedAt() < now - SECONDS_TO_RE_EVALUATE_ALL;
  }

  listSegmentIDs(feature: Feature): string[] {
    const mapIDs = new Set<string>();
    for (const rule of feature.getRulesList()) {
      for (const clause of rule.getClausesList()) {
        if (clause.getOperator() === Clause.Operator.SEGMENT) {
          clause.getValuesList().forEach((value) => mapIDs.add(value));
        }
      }
    }
    return Array.from(mapIDs);
  }

  assignUser(
    feature: Feature,
    user: User,
    segmentUsers: SegmentUser[],
    flagVariations: { [key: string]: string },
  ): [Reason, Variation] {
    for (const pf of feature.getPrerequisitesList()) {
      const variationId = flagVariations[pf.getFeatureId()];
      if (!variationId) {
        throw EVALUATOR_ERRORS.PrerequisiteVariationNotFound;
      }
      if (pf.getVariationId() !== variationId) {
        if (feature.getOffVariation() !== '') {
          const variation = this.findVariation(
            feature.getOffVariation(),
            feature.getVariationsList(),
          );
          const reason = createReason('', Reason.Type.PREREQUISITE);
          return [reason, variation];
        }
      }
    }
    // It doesn't assign the user in case of the feature is disabled and OffVariation is not set
    if (!feature.getEnabled() && feature.getOffVariation() !== '') {
      const variation = this.findVariation(feature.getOffVariation(), feature.getVariationsList());
      const reason = createReason('', Reason.Type.OFF_VARIATION);
      return [reason, variation];
    }

    // evaluate from top to bottom, return if one rule matches
    // evaluate targeting rules
    for (const target of feature.getTargetsList()) {
      if (target.getUsersList().includes(user.getId())) {
        const variation = this.findVariation(target.getVariation(), feature.getVariationsList());
        const reason = createReason('', Reason.Type.TARGET);
        return [reason, variation];
      }
    }

    // evaluate ruleset
    const rule = this.ruleEvaluator.evaluate(
      feature.getRulesList(),
      user,
      segmentUsers,
      flagVariations,
    );
    if (rule) {
      const strategy = rule.getStrategy();
      if (strategy == undefined) {
        throw EVALUATOR_ERRORS.StrategyNotFound;
      }
      const variation = this.strategyEvaluator.evaluate(
        strategy,
        user.getId(),
        feature.getVariationsList(),
        feature.getId(),
        feature.getSamplingSeed(),
      );
      const reason = createReason(rule.getId(), Reason.Type.RULE);
      return [reason, variation];
    }

    // use default strategy
    const defaultStrategy = feature.getDefaultStrategy();
    if (defaultStrategy == undefined) {
      throw EVALUATOR_ERRORS.DefaultStrategyNotFound;
    }

    const variation = this.strategyEvaluator.evaluate(
      defaultStrategy,
      user.getId(),
      feature.getVariationsList(),
      feature.getId(),
      feature.getSamplingSeed(),
    );
    const reason = createReason('', Reason.Type.DEFAULT);
    return [reason, variation];
  }

  getEvalFeatures(targetFeatures: Feature[], allFeatures: Feature[]): Feature[] {
    const allFeaturesMap = new Map<string, Feature>();
    for (const feature of allFeatures) {
      allFeaturesMap.set(feature.getId(), feature);
    }

    const evals1 = getFeaturesDependedOnTargets(targetFeatures, allFeaturesMap);
    const evals2 = getFeaturesDependsOnTargets(targetFeatures, allFeaturesMap);
    const mergedEvals = { ...evals1, ...evals2 };
    return Object.values(mergedEvals);
  }

  private findVariation(v: string, variations: Variation[]): Variation {
    const variation = variations.find((variation) => variation.getId() === v);
    if (!variation) {
      throw EVALUATOR_ERRORS.VariationNotFound;
    }
    return variation;
  }
}

enum Mark {
  Unvisited,
  Temporary,
  Permanently,
}

// FeatureIDsDependsOn returns the ids of the features that this feature depends on.
function getFeatureIDsDependsOn(feature: Feature): Array<string> {
  const ids: Array<string> = [];

  // Iterate over prerequisites and add their FeatureId
  feature.getPrerequisitesList().forEach((p) => {
    ids.push(p.getFeatureId());
  });

  // Iterate over rules and collect ids from clauses where the operator is FEATURE_FLAG
  feature.getRulesList().forEach((rule) => {
    rule.getClausesList().forEach((clause) => {
      if (clause.getOperator() === Clause.Operator.FEATURE_FLAG) {
        ids.push(clause.getAttribute());
      }
    });
  });

  return ids;
}

// Error types
class ErrCycleExists extends Error {
  constructor() {
    super('Cycle exists in the graph');
  }
}

class ErrFeatureNotFound extends Error {
  constructor() {
    super('Feature not found');
  }
}

// Topological sort function
function topologicalSort(features: Array<Feature>): Array<Feature> {
  const marks: { [key: string]: Mark } = {};
  const mapFeatures: { [key: string]: Feature } = {};

  features.forEach((f) => {
    marks[f.getId()] = Mark.Unvisited;
    mapFeatures[f.getId()] = f;
  });

  const sortedFeatures: Array<Feature> = [];

  const sort = (f: Feature) => {
    const fId = f.getId();
    if (marks[fId] === Mark.Permanently) return;

    if (marks[fId] === Mark.Temporary) {
      throw new ErrCycleExists();
    }

    marks[fId] = Mark.Temporary;

    const dependentFeatureIds = getFeatureIDsDependsOn(f);
    for (const fid of dependentFeatureIds) {
      const pf = mapFeatures[fid];
      if (!pf) {
        throw new ErrFeatureNotFound();
      }

      sort(pf);
    }

    marks[fId] = Mark.Permanently;
    sortedFeatures.push(f);
  };

  // Process each feature
  for (const f of features) {
    if (marks[f.getId()] === Mark.Unvisited) {
      sort(f);
    }
  }

  return sortedFeatures;
}

// getFeaturesDependedOnTargets returns the features that are depended on the target features.
// targetFeatures are included in the result.
function getFeaturesDependedOnTargets(
  targets: Array<Feature>,
  all: Map<string, Feature>,
): { [key: string]: Feature } {
  const evals: { [key: string]: Feature } = {};

  const dfs = (f: Feature): void => {
    // If the feature is already in the evals map, skip
    if (evals[f.getId()]) return;

    // Add feature to evals
    evals[f.getId()] = f;

    // Get dependent features recursively
    const featureDependencies = getFeatureIDsDependsOn(f);
    featureDependencies.forEach((fid) => {
      const target = all.get(fid);
      if (target !== undefined) {
        dfs(target);
      }
    });
  };

  // Start DFS for each target feature
  targets.forEach((f) => dfs(f));

  return evals;
}

// getFeaturesDependsOnTargets returns the features that depend on the target features.
// targetFeatures are included in the result.
function getFeaturesDependsOnTargets(
  targets: Array<Feature>,
  all: Map<string, Feature>,
): { [key: string]: Feature } {
  const evals: { [key: string]: Feature } = {};

  // Mark target features in the evals map initially
  targets.forEach((f) => {
    evals[f.getId()] = f;
  });

  // Depth-first search to determine if a feature depends on a target feature
  const dfs = (f: Feature): boolean => {
    if (evals[f.getId()]) {
      return true;
    }

    const featureDependencies = getFeatureIDsDependsOn(f);
    for (const fid of featureDependencies) {
      const dependentFeature = all.get(fid);
      if (dependentFeature && dfs(dependentFeature)) {
        // If the feature depends on one of the target features, add it to evals
        evals[f.getId()] = f;
        return true;
      }
    }
    return false;
  };

  // Apply DFS for all features, except target features
  all.forEach((f) => {
    dfs(f);
  });

  return evals;
}

function arrayToRecord(arr: Array<[string, string]>): Record<string, string> {
  return arr.reduce(
    (acc, [key, value]) => {
      acc[key] = value;
      return acc;
    },
    {} as Record<string, string>,
  );
}

export { Evaluator, getFeatureIDsDependsOn };

File: evaluation/go/evaluation.go
// Copyright 2026 The Bucketeer Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package evaluation

import (
	"encoding/json"
	"errors"
	"fmt"
	"sync"
	"time"

	"golang.org/x/exp/maps"
	"gopkg.in/yaml.v2"

	ftdomain "github.com/bucketeer-io/bucketeer/v2/pkg/feature/domain"
	ftproto "github.com/bucketeer-io/bucketeer/v2/proto/feature"
	userproto "github.com/bucketeer-io/bucketeer/v2/proto/user"
)

const (
	secondsToReEvaluateAll = 30 * 24 * 60 * 60 // 30 days
	secondsForAdjustment   = 10                // 10 seconds
)

var (
	ErrDefaultStrategyNotFound       = errors.New("evaluator: default strategy not found")
	ErrFeatureNotFound               = errors.New("evaluator: feature not found")
	ErrPrerequisiteVariationNotFound = errors.New("evaluator: prerequisite variation not found")
	ErrVariationNotFound             = errors.New("evaluator: variation not found")
	ErrUnsupportedStrategy           = errors.New("evaluator: unsupported strategy")
	ErrYAMLToJSONConversion          = errors.New("evaluator: failed to convert YAML to JSON")
)

func EvaluationID(featureID string, featureVersion int32, userID string) string {
	return fmt.Sprintf("%s:%d:%s", featureID, featureVersion, userID)
}

type evaluator struct {
	ruleEvaluator
	strategyEvaluator
	// variationCache caches YAML to JSON conversions using variation ID as the key.
	// Since variation IDs are UUIDs, they are globally unique and safe to use as cache keys.
	variationCache *sync.Map
}

func NewEvaluator() *evaluator {
	return &evaluator{
		variationCache: &sync.Map{},
	}
}

// This function will be removed once all the SDK clients are updated.
func (e *evaluator) EvaluateFeatures(
	fs []*ftproto.Feature,
	user *userproto.User,
	mapSegmentUsers map[string][]*ftproto.SegmentUser,
	targetTag string,
) (*ftproto.UserEvaluations, error) {
	return e.evaluate(fs, user, mapSegmentUsers, false, targetTag)
}

func (e *evaluator) EvaluateFeaturesByEvaluatedAt(
	fs []*ftproto.Feature,
	user *userproto.User,
	mapSegmentUsers map[string][]*ftproto.SegmentUser,
	prevUEID string,
	evaluatedAt int64,
	userAttributesUpdated bool,
	targetTag string,
) (*ftproto.UserEvaluations, error) {
	if prevUEID == "" {
		return e.evaluate(fs, user, mapSegmentUsers, true, targetTag)
	}
	now := time.Now()
	if evaluatedAt < now.Unix()-secondsToReEvaluateAll {
		return e.evaluate(fs, user, mapSegmentUsers, true, targetTag)
	}
	adjustedEvalAt := evaluatedAt - secondsForAdjustment
	updatedFeatures := make([]*ftproto.Feature, 0, len(fs))
	for _, feature := range fs {
		if feature.UpdatedAt > adjustedEvalAt {
			updatedFeatures = append(updatedFeatures, feature)
			continue
		}
		if userAttributesUpdated && len(feature.Rules) != 0 {
			updatedFeatures = append(updatedFeatures, feature)
		}
	}
	// If the UserEvaluationsID has changed, but both User Attributes and Feature Flags have not been updated,
	// it is considered unusual and a force update should be performed.
	if len(updatedFeatures) == 0 {
		return e.evaluate(fs, user, mapSegmentUsers, true, targetTag)
	}
	evalTargets, err := e.getEvalFeatures(updatedFeatures, fs)
	if err != nil {
		return nil, err
	}
	return e.evaluate(evalTargets, user, mapSegmentUsers, false, targetTag)
}

func (e *evaluator) evaluate(
	fs []*ftproto.Feature,
	user *userproto.User,
	mapSegmentUsers map[string][]*ftproto.SegmentUser,
	forceUpdate bool,
	targetTag string,
) (*ftproto.UserEvaluations, error) {

	flagVariations := map[string]string{}
	// fs need to be sorted in order from upstream to downstream.
	sortedFs, err := ftdomain.TopologicalSort(fs)
	if err != nil {
		return nil, err
	}
	evaluations := make([]*ftproto.Evaluation, 0, len(fs))
	archivedIDs := make([]string, 0)
	for _, feature := range sortedFs {
		var segmentUsers []*ftproto.SegmentUser
		for _, id := range e.ListSegmentIDs(feature) {
			segmentUsers = append(segmentUsers, mapSegmentUsers[id]...)
		}
		reason, variation, err := e.assignUser(feature, user, segmentUsers, flagVariations)
		if err != nil {
			return nil, err
		}
		// VariationId is used to check if prerequisite flag's result is what user expects it to be.
		// This must be set for ALL features (including archived) for dependency resolution to work
		flagVariations[feature.Id] = variation.Id

		if feature.Archived {
			// To keep response size small, the feature flags archived long time ago are excluded.
			if !e.isArchivedBeforeLastThirtyDays(feature) {
				archivedIDs = append(archivedIDs, feature.Id)
			}
			continue
		}
		// When the tag is set in the request,
		// it will return only the evaluations of flags that match the tag configured on the dashboard.
		// When empty, it will return all the evaluations of the flags in the environment.
		if targetTag != "" && !tagExist(feature.Tags, targetTag) {
			continue
		}
		// Convert YAML to JSON for client SDKs if needed
		// SDKs can retrieve variation values using the object variation interface.
		convertedValue := e.convertVariationValue(feature, variation)

		evaluationID := EvaluationID(feature.Id, feature.Version, user.Id)
		evaluation := &ftproto.Evaluation{
			Id:             evaluationID,
			FeatureId:      feature.Id,
			FeatureVersion: feature.Version,
			UserId:         user.Id,
			VariationId:    variation.Id,
			VariationName:  variation.Name,
			VariationValue: convertedValue,
			// Deprecated
			// FIXME: Remove the Variation when is no longer being used.
			// For security reasons, we should remove the variation description.
			// We copy the variation object to avoid race conditions when removing
			// the description directly from the `variation`
			Variation: &ftproto.Variation{
				Id:    variation.Id,
				Name:  variation.Name,
				Value: convertedValue,
			},
			Reason: reason,
		}
		evaluations = append(evaluations, evaluation)
	}
	// FIXME: Remove id once all SDKs will be updated.
	id := UserEvaluationsID(user.Id, user.Data, fs)
	userEvaluations := NewUserEvaluations(id, evaluations, archivedIDs, forceUpdate)
	return userEvaluations.UserEvaluations, nil
}

func tagExist(tags []string, target string) bool {
	for _, tag := range tags {
		if tag == target {
			return true
		}
	}
	return false
}

/*
IsArchivedBeforeLastThirtyDays returns a bool value
indicating whether the feature flag was archived within the last thirty days.
*/
func (e *evaluator) isArchivedBeforeLastThirtyDays(feature *ftproto.Feature) bool {
	if !feature.Archived {
		return false
	}
	now := time.Now()
	return feature.UpdatedAt < now.Unix()-secondsToReEvaluateAll
}

func (e *evaluator) ListSegmentIDs(feature *ftproto.Feature) []string {
	mapIDs := make(map[string]struct{})
	for _, r := range feature.Rules {
		for _, c := range r.Clauses {
			if c.Operator == ftproto.Clause_SEGMENT {
				for _, v := range c.Values {
					mapIDs[v] = struct{}{}
				}
			}
		}
	}
	ids := make([]string, 0, len(mapIDs))
	for id := range mapIDs {
		ids = append(ids, id)
	}
	return ids
}

func (e *evaluator) assignUser(
	feature *ftproto.Feature,
	user *userproto.User,
	segmentUsers []*ftproto.SegmentUser,
	flagVariations map[string]string,
) (*ftproto.Reason, *ftproto.Variation, error) {
	for _, pf := range feature.Prerequisites {
		variation, ok := flagVariations[pf.FeatureId]
		if !ok {
			return nil, nil, ErrPrerequisiteVariationNotFound
		}
		if pf.VariationId != variation {
			if feature.OffVariation != "" {
				variation, err := findVariation(feature.OffVariation, feature.Variations)
				return &ftproto.Reason{Type: ftproto.Reason_PREREQUISITE}, variation, err
			}
		}
	}
	// It doesn't assign the user in case of the feature is disabled and OffVariation is not set
	if !feature.Enabled && feature.OffVariation != "" {
		variation, err := findVariation(feature.OffVariation, feature.Variations)
		return &ftproto.Reason{Type: ftproto.Reason_OFF_VARIATION}, variation, err
	}
	// evaluate from top to bottom, return if one rule matches
	// evaluate targeting rules
	for i := range feature.Targets {
		if contains(user.Id, feature.Targets[i].Users) {
			variation, err := findVariation(feature.Targets[i].Variation, feature.Variations)
			return &ftproto.Reason{Type: ftproto.Reason_TARGET}, variation, err
		}
	}
	// evaluate ruleset
	rule, err := e.ruleEvaluator.Evaluate(feature.Rules, user, segmentUsers, flagVariations)
	if err != nil {
		return nil, nil, err
	}
	if rule != nil {
		variation, err := e.strategyEvaluator.Evaluate(
			rule.Strategy,
			user.Id,
			feature.Variations,
			feature.Id,
			feature.SamplingSeed,
		)
		return &ftproto.Reason{
			Type:   ftproto.Reason_RULE,
			RuleId: rule.Id,
		}, variation, err
	}
	// use default strategy
	if feature.DefaultStrategy == nil {
		return nil, nil, ErrDefaultStrategyNotFound
	}
	variation, err := e.strategyEvaluator.Evaluate(
		feature.DefaultStrategy,
		user.Id,
		feature.Variations,
		feature.Id,
		feature.SamplingSeed,
	)
	if err != nil {
		return nil, nil, err
	}
	return &ftproto.Reason{Type: ftproto.Reason_DEFAULT}, variation, nil
}

// GetPrerequisiteDownwards gets the features specified as prerequisite by the targetFeatures.
func (e *evaluator) GetPrerequisiteDownwards(
	targetFeatures, allFeatures []*ftproto.Feature,
) ([]*ftproto.Feature, error) {
	allFeaturesMap := make(map[string]*ftproto.Feature, len(allFeatures))
	for _, f := range allFeatures {
		allFeaturesMap[f.Id] = f
	}
	return maps.Values(ftdomain.GetFeaturesDependedOnTargets(targetFeatures, allFeaturesMap)), nil
}

func (e *evaluator) getEvalFeatures(
	targetFeatures, allFeatures []*ftproto.Feature,
) ([]*ftproto.Feature, error) {
	all := make(map[string]*ftproto.Feature, len(allFeatures))
	for _, f := range allFeatures {
		all[f.Id] = f
	}

	evals1 := ftdomain.GetFeaturesDependedOnTargets(targetFeatures, all)
	evals2 := ftdomain.GetFeaturesDependsOnTargets(targetFeatures, all)
	evals := e.mapMerge(evals1, evals2)
	return maps.Values(evals), nil
}

func (e *evaluator) mapMerge(m1, m2 map[string]*ftproto.Feature) map[string]*ftproto.Feature {
	for k, v := range m2 {
		m1[k] = v
	}
	return m1
}

func findVariation(v string, vs []*ftproto.Variation) (*ftproto.Variation, error) {
	for i := range vs {
		if vs[i].Id == v {
			return vs[i], nil
		}
	}
	return nil, ErrVariationNotFound
}

func contains(needle string, haystack []string) bool {
	for i := range haystack {
		if haystack[i] == needle {
			return true
		}
	}
	return false
}

// convertVariationValue converts YAML to JSON if needed, with in-memory caching.
// This ensures client SDKs can retrieve variation values using the object variation interface.
// Only performs conversion when the variation type is YAML.
// Cache key uses feature.UpdatedAt + variation.Id to ensure cache invalidation when variations are updated.
func (e *evaluator) convertVariationValue(
	feature *ftproto.Feature,
	variation *ftproto.Variation,
) string {
	// Only convert if type is YAML
	if feature.VariationType != ftproto.Feature_YAML {
		return variation.Value
	}

	// Cache key: {featureUpdatedAt}:{variationId}
	// This ensures cache is invalidated when the feature (and its variations) are updated,
	// including changes from auto operations that don't increment feature.Version
	cacheKey := fmt.Sprintf("%d:%s", feature.UpdatedAt, variation.Id)

	// Check cache first
	if cached, ok := e.variationCache.Load(cacheKey); ok {
		return cached.(string)
	}

	// Convert YAML to JSON
	jsonValue, err := yamlToJSON(variation.Value)
	if err != nil {
		// Log would be helpful here, but to avoid dependency injection,
		// we return the original value as a fallback
		return variation.Value
	}

	// Cache the result for future requests
	e.variationCache.Store(cacheKey, jsonValue)
	return jsonValue
}

// yamlToJSON converts a YAML string to a JSON string.
func yamlToJSON(yamlStr string) (string, error) {
	var data interface{}
	if err := yaml.Unmarshal([]byte(yamlStr), &data); err != nil {
		return "", fmt.Errorf("%w: %v", ErrYAMLToJSONConversion, err)
	}

	// Convert map[interface{}]interface{} to map[string]interface{} for JSON compatibility
	data = convertMapKeys(data)

	jsonBytes, err := json.Marshal(data)
	if err != nil {
		return "", fmt.Errorf("%w: %v", ErrYAMLToJSONConversion, err)
	}
	return string(jsonBytes), nil
}

// convertMapKeys recursively converts map[interface{}]interface{} to map[string]interface{}
// This is necessary because yaml.v2 unmarshals to map[interface{}]interface{},
// but json.Marshal requires map[string]interface{}.
func convertMapKeys(input interface{}) interface{} {
	switch x := input.(type) {
	case map[interface{}]interface{}:
		m := make(map[string]interface{})
		for k, v := range x {
			m[fmt.Sprintf("%v", k)] = convertMapKeys(v)
		}
		return m
	case []interface{}:
		for i, v := range x {
			x[i] = convertMapKeys(v)
		}
	}
	return input
}
